### PR TITLE
Add warnings when graphing unaggregated data or truncated rows

### LINKED ISSUE #626: when a user runs a query that returns too many rows and is truncated, if the user graphs the results, we should let them know not all results came back

### LINKED ISSUE #2475: The wording on the table view about data being truncated is unclear
It's currently a little ambiguous when a user runs a query and the results are truncated.  The messaging reads ... `Showing max of 2000 rows` ... which it sounds like some users are interpreting as just a generalized warning that Metabase will only show 2k rows, but it's not clearly informing the user that their current query has hit the 2k row limit and is being truncated.

The end result of this is that users don't see all the data they expect and are confused about why.

### LINKED ISSUE #2708: Should we prevent or warn people when graphing "raw data"?
Someone asked me why they couldn't make a multiseries out of two cards with the same breakouts, and it turned out the cards were something like `raw data, by created_at and total`, which happens to graph basically the same as `sum of total by created_at` (assuming the raw data query isn't truncated to 2000 rows) due to the way our dc.js graphing code works (it basically sums up rows where the X value is the same), but it's clearly not the right way to get that result. It's also not great because we only show the "Showing max of 2000 rows" warning when showing the table.

Could we somehow warn or otherwise push users in the right direction when they do something like that?

### LINKED ISSUE #3688: When summing by month, the NULL group's sum is added to a random month
### Bugs

If you're filing an issue about a bug please include as much information
as you can including the following.
- Chrome 53
- macOS 10.12.1
- Data is in PostgreSQL 9.6
- Metabase 0.20.3
- Hosted via Docker
- Metabase internal database is PostgreSQL 9.5

Don't have time to create a full repro, but my query is _Count of Rows by related table -> timestamp: month_ and the CSV export looks like this:

``` csv
timestamp,count
2014-03-01T00:00:00.000+01:00,0
2014-04-01T00:00:00.000+02:00,1
2014-05-01T00:00:00.000+02:00,3
2014-06-01T00:00:00.000+02:00,7
2014-07-01T00:00:00.000+02:00,10
2014-08-01T00:00:00.000+02:00,9
2014-09-01T00:00:00.000+02:00,10
2014-10-01T00:00:00.000+02:00,12
2014-11-01T00:00:00.000+01:00,17
2014-12-01T00:00:00.000+01:00,17
2015-01-01T00:00:00.000+01:00,27
2015-02-01T00:00:00.000+01:00,28
2015-03-01T00:00:00.000+01:00,50
2015-04-01T00:00:00.000+02:00,57
2015-05-01T00:00:00.000+02:00,74
2015-06-01T00:00:00.000+02:00,89
2015-07-01T00:00:00.000+02:00,91
2015-08-01T00:00:00.000+02:00,129
2015-09-01T00:00:00.000+02:00,157
2015-10-01T00:00:00.000+02:00,189
2015-11-01T00:00:00.000+01:00,150
2015-12-01T00:00:00.000+01:00,164
2016-01-01T00:00:00.000+01:00,265
2016-02-01T00:00:00.000+01:00,266
2016-03-01T00:00:00.000+01:00,294
2016-04-01T00:00:00.000+02:00,296
2016-05-01T00:00:00.000+02:00,332
2016-06-01T00:00:00.000+02:00,429
2016-07-01T00:00:00.000+02:00,417
2016-08-01T00:00:00.000+02:00,465
2016-09-01T00:00:00.000+02:00,461
2016-10-01T00:00:00.000+02:00,399
,358
```

And the line chart looks like this:

The 2015-11 peak says the value is 508 when hovering over it. The keen-eyed will notice that this is `2015-11 sum + NULL group sum` (`150 + 358`).

Interestingly, the correct value is displayed in the pie chart and table modes.

### PR DESCRIPTION
**Pull request by tlrobinson**
* [x] Resolves #626 (when a user runs a query that returns too many rows and is truncated, if the user graphs the results, we should let them know not all results came back)
* [x] Resolves #2708 (Should we prevent or warn people when graphing "raw data"?)
* [x] Resolves #3688 (When summing by month, the NULL group's sum is added to a random month)
* [x] Resolves #2475 (The wording on the table view about data being truncated is unclear)
* ~~#2913 (Points with the same X value are being added together in the line graph )~~

### CODE REVIEW

**Comment by tlrobinson** [comment]
screen shot 2016-11-15 at 5 51 58 pm
screen shot 2016-11-15 at 5 51 51 pm

**Comment by tlrobinson** [comment]
@metabase/core-developers This is ready for review. The warning copy might need some work:

* `Data truncated to N rows.`
* `Data includes an unaggregated dimension.`
* `Data includes missing dimension values.`

File: frontend/src/metabase/query_builder/components/QueryVisualization.jsx [hunk 1]
```diff
@@ -95,37 +98,25 @@ export default class QueryVisualization extends Component {
                         cancelFn={this.props.cancelQueryFn}
                     />
                 </div>
-                <div className="absolute right z4 flex align-center">
+                <div className="absolute right z4 flex align-center"  style={{ lineHeight: 0 /* needed to align icons :-/ */ }}>
```
Comment (kdoh) [inline frontend/src/metabase/query_builder/components/QueryVisualization.jsx]: This is a bit of a bummer. I wonder if there's a way around it.

File: frontend/src/metabase/visualizations/components/Visualization.jsx [hunk 2]
```diff
@@ -84,35 +88,60 @@ export default class Visualization extends Component {
...
+        this.updateWarnings();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (!Utils.equals(this.getWarnings(prevProps, prevState), this.getWarnings())) {
+            this.updateWarnings();
+        }
+    }
+
+    getWarnings(props = this.props, state = this.state) {
```
Comment (kdoh) [inline frontend/src/metabase/visualizations/components/Visualization.jsx]: I'm assuming you did this to provide a nicer shorthand for props and state inside the function? I'm a fan. I did notice you didn't use the same pattern in some of the other non lifecycle component methods so I'm curious if there was a particular reason it was useful here.

**Comment by kdoh** [comment]
Left a couple comments. Overall LGTM.

**Comment by tlrobinson** [comment]
@metabase/core-developers Anyone have any opinions on the warning message copy?

**Comment by mazameli** [comment]
> Anyone have any opinions on the warning message copy?

For the unaggregated warning, can we be specific about the field by any chance? This is way too verbose probably, but it's a tricky concept to explain, and tricky to explain what we do about it: 

`"Tax" is an unaggregated field: if it has more than one value at a point on the x-axis, the other values will be ignored.` 

Maybe instead:
`"Tax" is an unaggregated field. [Learn more](url)`, which would require a page / anchor link to explain what we do in this case.

**Comment by tlrobinson** [comment]
Should be able to do something like showing the unaggregated column, but we can't really have a clickable link in a tooltip...

**Comment by tlrobinson** [comment]
This should be ready to go if anyone wants to take another look.

**Comment by tlrobinson** [comment]
@metabase/core-developers I'd like to GTM'd for 0.21, anyone want to review it?

**Comment by mazameli** [comment]
Taking a look right now. :D

**Comment by mazameli** [comment]
screen shot 2016-11-28 at 10 53 37 am

There are a few things that look a little off to me with this content block. I know the icon and message were there before this branch, so maybe this is out of scope, but:
- Let's make the paragraph text darker (like our normal paragraph text color). Then the text could read, "Which fields do you want to use for your x and y axes?" And the button text could read, "Choose fields."
- The button looks tiny. I think we should just use a regular primary button.
- Let's scrap the warning icon. With the other changes and the removal of this icon, the interaction will feel less like an error state, in a good way.

screen shot 2016-11-28 at 10 53 59 am
- I think I'll get you a slightly modified version of this icon to replace this one. I think the corners could be a bit more rounded, and the exclamation mark could be a little bolder so it's discernible at small sizes.

screen shot 2016-11-28 at 10 56 06 am
- Looking at this now, this particular warning in the chart settings feels like it should actually be placed next to the dropdown with the field that's prompting the warning. Would that work? In the chart settings, are we displaying any warnings except for the field-based ones? (It seemed from what I could tell that we weren't displaying the 2,000-rows-only warning in the chart settings modal).
- In the tooltip, I see that you're using the warning icon as a bullet for the `<ul>`. Interesting idea, but I think it's too much, especially when there's only a single warning. I think we can get away with no bullets, and just having sufficient margin-bottom after each list item.

**Comment by mazameli** [comment]
Here's an SVG path for a new warning triangle:

```M12.3069589,4.52260192 C14.3462632,1.2440969 17.653446,1.24541073 19.691933,4.52260192 L31.2249413,23.0637415 C33.2642456,26.3422466 31.7889628,29 27.9115531,29 L4.08733885,29 C0.218100769,29 -1.26453645,26.3409327 0.77395061,23.0637415 L12.3069589,4.52260192 Z M18.0499318,23.0163223 C18.0499772,23.0222378 18.05,23.0281606 18.05,23.0340907 C18.05,23.3266209 17.9947172,23.6030345 17.8840476,23.8612637 C17.7737568,24.1186089 17.6195847,24.3426723 17.4224081,24.5316332 C17.2266259,24.7192578 16.998292,24.8660439 16.7389806,24.9713892 C16.4782454,25.0773129 16.1979962,25.1301134 15.9,25.1301134 C15.5950083,25.1301134 15.3111795,25.0774024 15.0502239,24.9713892 C14.7901813,24.8657469 14.5629613,24.7183609 14.3703047,24.5298034 C14.177545,24.3411449 14.0258626,24.1177208 13.9159524,23.8612637 C13.8052827,23.6030345 13.75,23.3266209 13.75,23.0340907 C13.75,22.7411889 13.8054281,22.4661013 13.9165299,22.2109786 C14.0264627,21.9585404 14.1779817,21.7374046 14.3703047,21.5491736 C14.5621821,21.3613786 14.7883231,21.2126553 15.047143,21.1034656 C15.3089445,20.9930181 15.593871,20.938068 15.9,20.938068 C16.1991423,20.938068 16.4804862,20.9931136 16.7420615,21.1034656 C17.0001525,21.2123478 17.2274115,21.360472 17.4224081,21.5473437 C17.6191428,21.7358811 17.7731504,21.957652 17.88347,22.2109786 C17.9124619,22.2775526 17.9376628,22.3454862 17.9590769,22.414741 C18.0181943,22.5998533 18.05,22.7963729 18.05,23 C18.05,23.0054459 18.0499772,23.0108867 18.0499318,23.0163223 L18.0499318,23.0163223 Z M17.7477272,14.1749999 L17.7477272,8.75 L14.1170454,8.75 L14.1170454,14.1749999 C14.1170454,14.8471841 14.1572355,15.5139742 14.2376219,16.1753351 C14.3174838,16.8323805 14.4227217,17.5019113 14.5533248,18.1839498 L14.5921937,18.3869317 L17.272579,18.3869317 L17.3114479,18.1839498 C17.442051,17.5019113 17.5472889,16.8323805 17.6271507,16.1753351 C17.7075371,15.5139742 17.7477272,14.8471841 17.7477272,14.1749999 Z```

**Pull request merged**